packages/storage-plus/src/snapshot.rs

```diff
@@ -0,0 +1,371 @@
+                    let start = Bound::inclusive(U64Key::from(height));
+                    let first = self
+                        .changelog
+                        .prefix(k.clone())
+                        .range(store, Some(start), None, Order::Ascending)
+                        .next()
+                        .transpose()?;
+                    if first.is_none() {
+                        // there must be at least one open checkpoint and no changelog for the given address since then
+                        return Ok(true);
+                    }
+                }
+                // otherwise, we don't save this
+                Ok(false)
+            }
+        }
+    }
+
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
+        let old = self.may_load(store, k.clone())?;
+        self.changelog
+            .save(store, (k, U64Key::from(height)), &ChangeSet { old })
+    }
+
+    pub fn save(&self, store: &mut dyn Storage, k: K, data: &T, height: u64) -> StdResult<()> {
+        if self.should_checkpoint(store, &k)? {
+            self.write_change(store, k.clone(), height)?;
+        }
+        self.primary.save(store, k, data)
+    }
+
+    pub fn remove(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
+        if self.should_checkpoint(store, &k)? {
+            self.write_change(store, k.clone(), height)?;
+        }
+        self.primary.remove(store, k);
+        Ok(())
+    }
+
+    /// load will return an error if no data is set at the given key, or on parse error
+    pub fn load(&self, store: &dyn Storage, k: K) -> StdResult<T> {
+        self.primary.load(store, k)
+    }
+
+    /// may_load will parse the data stored at the key if present, returns Ok(None) if no data there.
+    /// returns an error on issues parsing
+    pub fn may_load(&self, store: &dyn Storage, k: K) -> StdResult<Option<T>> {
+        self.primary.may_load(store, k)
+    }
+
+    // may_load_at_height reads historical data from given checkpoints.
+    // only guaranteed to give correct data if Strategy::EveryBlock or
+    // Strategy::Selected and h is registered as checkpoint
+    pub fn may_load_at_height(
+        &self,
+        store: &dyn Storage,
+        k: K,
+        height: u64,
+    ) -> StdResult<Option<T>> {
+        // this will look for the first snapshot of the given address >= given height
+        // If None, there is no snapshot since that time.
+        let start = Bound::inclusive(U64Key::new(height));
+        let first = self
+            .changelog
+            .prefix(k.clone())
+            .range(store, Some(start), None, Order::Ascending)
+            .next();
+
+        if let Some(r) = first {
+            // if we found a match, return this last one
+            r.map(|(_, v)| v.old)
+        } else {
+            // otherwise, return current value
+            self.may_load(store, k)
```

Comment:
What about signalling / exposing this difference? Particularly for `Selected`, it can be useful. At least for debugging.
Maybe by providing an alternative method, `is_correct()` or so, not to spoil the return value here.

Reply:
When the data is not available, I can either return possibly incorrect data, or return an error.

I think you are correct that an error is better.

That applies to any historical query for `Selected::Never`

Reply:
I addressed this, it down returns `StdError::NotFound` if there is no relevant checkpoint

Reply:
It is my understanding that it may apply to `Selected` as well: If you're using `Selected`, and didn't create / request a checkpoint for that specific height, you cannot tell for sure if the returned value is correct. It _may_ be correct, but it also be an updated value, which didn't get snapshotted at that height.

Reply:
What I'm proposing is returning the value nevertheless, but return also a flag (or, maybe better, provide another method) to indicate if the value is guaranteed to be correct for that height, or not.

Reply:
I've seen you have implemented returning an error when the value is not guaranteed to be correct.

I prefer it, as it's less error prone: if you are using `Selected` and didn't get the guaranteed values for your heights, you're probably using it wrong. Same if you're not using snapshots at all, and pretend to get some historical values.